### ISSUE TITLE
Pipelines should be able to publish nodejs packages to NPM

### ISSUE THREAD

**Issue by feoh** [mitodl/ol-infrastructure#2190]
### Description/Context

Right now we have no way of allowing pipelines to publish to NPM.

This isn't great, and forces devs to build and publish manually on their own machines.

### Plan/Design

We could use this [concourse resource](https://github.com/timotto/concourse-npm-resource) to accomplish the task.

We'd just need to plumb the appropriate token through Vault.

Completion of this task should include a sample pipeline , or maybe even just finish off the NPM publishing piece of https://github.com/mitodl/ol-infrastructure/blob/main/src/ol_concourse/pipelines/libraries/open_api_clients.py

**Comment by feoh** [comment]
Adding the semver resource: https://github.com/concourse/semver-resource because the current bash scripts in the create-release tasks just go boom in the Concourse context.

**Comment by feoh** [comment]
I am a bit at a loss with what to do / where to go with this project.

The npm-package resource is busted. It seems to want a typescript compiler on the command line which node:alpine doesn't provide. So I guess we'd need to fork that project and update/ fix the bug. I saw some comments on other projects citing node/npm 6 -> 7 difference.

Then there's concourse-semver. Turns out my attempt at using it are totally inappropriate, as we can't do our version bumping in Git at all.

So we'd need to figure out how to construct an appropriate environment within a concourse container to run the bumpver Python script. That seems doable at least.

Bottom line, there's still a sizable amount of work left here to make this pipeline usable, and even then I feel like we'd need to get commitment from a developer resource to work in lock step with us to ensure that what we're building actually makes sense.

**Comment by feoh** [comment]
I've destroyed the open-api-clients pipeline for now to avoid it causing any further damage. 

My branch for this work is [here](https://github.com/mitodl/ol-infrastructure/tree/cpatti_add_npm_publish) but this pipeline shouldn't be built and run without extensive modifications as it will errantly push new versions to mit-open which isn't wnat we want.

**Comment by feoh** [comment]
concourse-npm-resource is dramatically limited so we can't use it. Dropped back to npm publish shell script.

That script is having [issues](https://cicd.odl.mit.edu/teams/infrastructure/pipelines/open-api-clients/jobs/publish/builds/14).

We are breaths away from the finish line. I can FEEL it :)

**Comment by feoh** [comment]
I am hard blocked on [this](https://cicd.odl.mit.edu/teams/infrastructure/pipelines/open-api-clients/jobs/publish/builds/23) build  failure.

I see a bunch of github issues floating around saying that yarn doesn't support setting an NPM token as an environment variable.

I tried:

NPM_TOKEN
NPM_AUTH_TOKEN
NODE_AUTH_TOKEN

As I've seen refs to all three, but none actually worked.

I suspect we could cram the token into .npmrc but I'm not sure what if any deleterious effects that could have on this package, so this will likely need to wait until Nathan comes back from vacation unless someone stalwart is willing to take a look :)

**Comment by feoh** [comment]
Re-opening due to reports of this [failed build](https://cicd.odl.mit.edu/teams/infrastructure/pipelines/open-api-clients/jobs/publish/builds/56).

We're not actually generating any commits after version bump. My PutStep just does the moral equivalent of `git push`.

**Comment by feoh** [comment]
Re-fixed by @blarghmatey - thanks!

https://cicd.odl.mit.edu/teams/infrastructure/pipelines/open-api-clients/jobs/publish/builds/58
